Adding pydot --fixture-graph

fruch opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pytest-dev/pytest-fixture-tools/3)
<!-- Reviewable:end -->

fruch commented:
@bubenkoff @amakhnach I've did some changes to CI, and also remove the usage of tox (i.e. not much need in Travis anymore)

I think this one is very helpful (it was to me), would be honored to have it in.

bubenkoff commented:
Review status: 0 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[pytest_fixture_tools/plugin.py, line 141 at r2](https://reviewable.io:443/reviews/pytest-dev/pytest-fixture-tools/3#-L-yXGQPzZoB0f1yB0ta:-L-yXGQQLy0rlCW_ioR8:br8rjov) ([raw file](https://github.com/pytest-dev/pytest-fixture-tools/blob/81f96a5404ccc8699875aa86abe2f56c21efe881/pytest_fixture_tools/plugin.py#L141)):*
> LOG_DEST_DIR
This is both undocumented and not so flexible - I would then make it a command line argument so --fixture-graph=<file name>

---

*[pytest_fixture_tools/plugin.py, line 147 at r2](https://reviewable.io:443/reviews/pytest-dev/pytest-fixture-tools/3#-L-yXaC2Us3ZG5Jg8fQl:-L-yXaC2Us3ZG5Jg8fQm:b-nmz98y) ([raw file](https://github.com/pytest-dev/pytest-fixture-tools/blob/81f96a5404ccc8699875aa86abe2f56c21efe881/pytest_fixture_tools/plugin.py#L147)):*
> ```Python
>         tw.line()
>         tw.sep("-", "fixture-graph")
>         tw.line("created at {}.dot.".format(filename))
> ```

how about to directly generate a png file, so people will not have to run dot manually?
also need to document the optional dependency of course

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pytest-dev/pytest-fixture-tools/3)*
<!-- Sent from Reviewable.io -->

fruch commented:
Review status: 0 of 7 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pytest_fixture_tools/plugin.py, line 141 at r2](https://reviewable.io:443/reviews/pytest-dev/pytest-fixture-tools/3#-L-yXGQPzZoB0f1yB0ta:-L-zJhLY73aHZOzo56gj:b-mhx5i5) ([raw file](https://github.com/pytest-dev/pytest-fixture-tools/blob/81f96a5404ccc8699875aa86abe2f56c21efe881/pytest_fixture_tools/plugin.py#L141)):*
<details><summary><i>Previously, bubenkoff (Anatoly Bubenkov) wrote…</i></summary><blockquote>

> LOG_DEST_DIR
This is both undocumented and not so flexible - I would then make it a command line argument so --fixture-graph=<file name>
</blockquote></details>

I've removed this one, and added a new command line param for the output-dir

---

*[pytest_fixture_tools/plugin.py, line 147 at r2](https://reviewable.io:443/reviews/pytest-dev/pytest-fixture-tools/3#-L-yXaC2Us3ZG5Jg8fQl:-L-zJX88UKBPgW4xTMG6:b6kh244) ([raw file](https://github.com/pytest-dev/pytest-fixture-tools/blob/81f96a5404ccc8699875aa86abe2f56c21efe881/pytest_fixture_tools/plugin.py#L147)):*
<details><summary><i>Previously, bubenkoff (Anatoly Bubenkov) wrote…</i></summary><blockquote>

how about to directly generate a png file, so people will not have to run dot manually?
also need to document the optional dependency of course
</blockquote></details>

I've updated it, hopfully it's clearer now

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pytest-dev/pytest-fixture-tools/3)*
<!-- Sent from Reviewable.io -->

bubenkoff commented:
Review status: 0 of 7 files reviewed at latest revision, 2 unresolved discussions.

---

*[pytest_fixture_tools/plugin.py, line 39 at r3](https://reviewable.io:443/reviews/pytest-dev/pytest-fixture-tools/3#-L0-8E4zuWiyxdZz-ju_:-L0-8E4zuWiyxdZz-jua:biviaqs) ([raw file](https://github.com/pytest-dev/pytest-fixture-tools/blob/a232b7c9e7724725abfb890de904deef38f3722d/pytest_fixture_tools/plugin.py#L39)):*
> ```Python
>                     action="store_true", dest="fixture_graph", default=False,
>                     help="create .dot fixture graph for each test")
>     group.addoption('--fixture-graph-output-dir',
> ```

please document in the readme

---

*[pytest_fixture_tools/plugin.py, line 42 at r3](https://reviewable.io:443/reviews/pytest-dev/pytest-fixture-tools/3#-L0-8GX1H_HD8raq1laY:-L0-8GX1H_HD8raq1laZ:biviaqs) ([raw file](https://github.com/pytest-dev/pytest-fixture-tools/blob/a232b7c9e7724725abfb890de904deef38f3722d/pytest_fixture_tools/plugin.py#L42)):*
> ```Python
>                     action="store_true", dest="fixture_graph_output_dir", default="artifacts",
>                     help="select the location for the output of fixture graph. defaults to 'artifacts'")
>     group.addoption('--fixture-graph-output-type',
> ```

please document in the readme

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pytest-dev/pytest-fixture-tools/3)*
<!-- Sent from Reviewable.io -->

fruch commented:
Review status: 0 of 7 files reviewed at latest revision, 2 unresolved discussions.

---

*[pytest_fixture_tools/plugin.py, line 39 at r3](https://reviewable.io:443/reviews/pytest-dev/pytest-fixture-tools/3#-L0-8E4zuWiyxdZz-ju_:-L00fEVrioBuEkwRK6EV:b-896fix) ([raw file](https://github.com/pytest-dev/pytest-fixture-tools/blob/a232b7c9e7724725abfb890de904deef38f3722d/pytest_fixture_tools/plugin.py#L39)):*
<details><summary><i>Previously, bubenkoff (Anatoly Bubenkov) wrote…</i></summary><blockquote>

please document in the readme
</blockquote></details>

Done.

---

*[pytest_fixture_tools/plugin.py, line 42 at r3](https://reviewable.io:443/reviews/pytest-dev/pytest-fixture-tools/3#-L0-8GX1H_HD8raq1laY:-L00fFjF59ozroDjgw2L:b-896fix) ([raw file](https://github.com/pytest-dev/pytest-fixture-tools/blob/a232b7c9e7724725abfb890de904deef38f3722d/pytest_fixture_tools/plugin.py#L42)):*
<details><summary><i>Previously, bubenkoff (Anatoly Bubenkov) wrote…</i></summary><blockquote>

please document in the readme
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pytest-dev/pytest-fixture-tools/3)*
<!-- Sent from Reviewable.io -->

bubenkoff commented:
Reviewed 3 of 5 files at r1, 3 of 4 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pytest-dev/pytest-fixture-tools/3)*
<!-- Sent from Reviewable.io -->

bubenkoff commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pytest-dev/pytest-fixture-tools/3#-:-L01TqVzIuY7YvyAeCB8:bnfp4nl)*
<!-- Sent from Reviewable.io -->

mikeage commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pytest_fixture_tools/plugin.py, line 131 at r4](https://reviewable.io:443/reviews/pytest-dev/pytest-fixture-tools/3#-L02vfUm6BC1bi0ilqYC:-L02vfUm6BC1bi0ilqYD:bsvbpm0) ([raw file](https://github.com/pytest-dev/pytest-fixture-tools/blob/02fa7fd4553d2341117d445f6828db44a5d80f50/pytest_fixture_tools/plugin.py#L131)):*
> ```Python
> 
>             loc = getlocation(fixture_data[0].func, curdir)
>             if 'pytest_vgw' in loc:
> ```

I'm pretty sure this doesn't belong here

---

*[pytest_fixture_tools/plugin.py, line 158 at r4](https://reviewable.io:443/reviews/pytest-dev/pytest-fixture-tools/3#-L02vpSMeni142eAOls8:-L02vpSMeni142eAOls9:b-x66zmq) ([raw file](https://github.com/pytest-dev/pytest-fixture-tools/blob/02fa7fd4553d2341117d445f6828db44a5d80f50/pytest_fixture_tools/plugin.py#L158)):*
> ```Python
>             tw.line("created {}.{}.".format(filename, output_type))
>         except Exception:
>             tw.line("grpahvis wasn't found in PATH")
> ```

graphviz

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pytest-dev/pytest-fixture-tools/3)*
<!-- Sent from Reviewable.io -->

fruch commented:
Review status: 6 of 7 files reviewed at latest revision, 2 unresolved discussions.

---

*[pytest_fixture_tools/plugin.py, line 131 at r4](https://reviewable.io:443/reviews/pytest-dev/pytest-fixture-tools/3#-L02vfUm6BC1bi0ilqYC:-L035xmrUBqs_yKXSbOT:b-896fix) ([raw file](https://github.com/pytest-dev/pytest-fixture-tools/blob/02fa7fd4553d2341117d445f6828db44a5d80f50/pytest_fixture_tools/plugin.py#L131)):*
<details><summary><i>Previously, mikeage (Mike Miller) wrote…</i></summary><blockquote>

I'm pretty sure this doesn't belong here
</blockquote></details>

Done.

---

*[pytest_fixture_tools/plugin.py, line 158 at r4](https://reviewable.io:443/reviews/pytest-dev/pytest-fixture-tools/3#-L02vpSMeni142eAOls8:-L035ygapl9h0mQ3xFZ7:b-896fix) ([raw file](https://github.com/pytest-dev/pytest-fixture-tools/blob/02fa7fd4553d2341117d445f6828db44a5d80f50/pytest_fixture_tools/plugin.py#L158)):*
<details><summary><i>Previously, mikeage (Mike Miller) wrote…</i></summary><blockquote>

graphviz
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pytest-dev/pytest-fixture-tools/3)*
<!-- Sent from Reviewable.io -->

fruch commented:
@bubenkoff you gave LGTM on it ? anything else is missing ? is it waiting for @mikeage to approve his comments ? (this reviewable is a bit confusing)

amakhnach commented on tests/test_fixtrue_graph.py:
```diff
@@ -0,0 +1,28 @@
+
+
+def test_fixture_graph_created(testdir):
+    """Check that --show-fixture-duplicates will give us list of duplicates."""
```
Copy-pasted docstring.

fruch commented:
oops, lazy me. will fix it tomorrow once I'm at my laptop again.

amakhnach commented:
LGTM. Merging.
